### PR TITLE
[SPARK-6130] [SQL] support if not exists for insert overwrite into partition in hiveQl

### PR DESCRIPTION
**Pull request by adrian-wang**
Standard syntax:
INSERT OVERWRITE TABLE tablename1 [PARTITION (partcol1=val1, partcol2=val2 ...) [IF NOT EXISTS]] select_statement1 FROM from_statement;
INSERT INTO TABLE tablename1 [PARTITION (partcol1=val1, partcol2=val2 ...)] select_statement1 FROM from_statement;
 
Hive extension (multiple inserts):
FROM from_statement
INSERT OVERWRITE TABLE tablename1 [PARTITION (partcol1=val1, partcol2=val2 ...) [IF NOT EXISTS]] select_statement1
[INSERT OVERWRITE TABLE tablename2 [PARTITION ... [IF NOT EXISTS]] select_statement2]
[INSERT INTO TABLE tablename2 [PARTITION ...] select_statement2] ...;
FROM from_statement
INSERT INTO TABLE tablename1 [PARTITION (partcol1=val1, partcol2=val2 ...)] select_statement1
[INSERT INTO TABLE tablename2 [PARTITION ...] select_statement2]
[INSERT OVERWRITE TABLE tablename2 [PARTITION ... [IF NOT EXISTS]] select_statement2] ...;
 
Hive extension (dynamic partition inserts):
INSERT OVERWRITE TABLE tablename PARTITION (partcol1[=val1], partcol2[=val2] ...) select_statement FROM from_statement;
INSERT INTO TABLE tablename PARTITION (partcol1[=val1], partcol2[=val2] ...) select_statement FROM from_statement;

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala [hunk 1]
```diff
@@ -1103,7 +1103,7 @@ class DataFrame protected[sql](
   @Experimental
   def insertInto(tableName: String, overwrite: Boolean): Unit = {
     sqlContext.executePlan(InsertIntoTable(UnresolvedRelation(Seq(tableName)),
-      Map.empty, logicalPlan, overwrite)).toRdd
+      Map.empty, logicalPlan, overwrite, false)).toRdd
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala]: Use named parameters for all boolean args please.

**Comment by marmbrus** [comment]
Sorry for the delay.  Only a minor comment.  Can you please update?

**Comment by adrian-wang** [comment]
The updated PR test of 29646 has passed...
29645 is an invalid rebase, let's just ignore it.

**Comment by adrian-wang** [comment]
retest this please.

**Comment by adrian-wang** [comment]
ping @marmbrus

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala [hunk 2]
```diff
@@ -125,7 +125,8 @@ case class InsertIntoTable(
     table: LogicalPlan,
     partition: Map[String, Option[String]],
     child: LogicalPlan,
-    overwrite: Boolean)
+    overwrite: Boolean,
+    ifNotExists: Boolean)
```
Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Seems its not valid to have both be true.  Can we add an assertion?

**Comment by marmbrus** [comment]
Sorry for letting this go stale, can you update?  One minor comment otherwise LGTM.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala (hunk 2 above)
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicOperators.scala]: Actually ifNotExists can only be true when overwrite is true... It seems confusing but really, you can check the description of this pr.

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

**Pull request closed without merging**